bluesky/plan_patterns.py

```diff
@@ -0,0 +1,213 @@
+        y center
+    x_range : float
+        x width of spiral
+    y_range : float
+        y width of spiral
+    dr : float
+        delta radius
+    factor : float
+        radius gets divided by this
+    per_step : callable, optional
+        hook for cutomizing action of inner loop (messages per step)
+        See docstring of bluesky.plans.one_nd_step (the default) for
+        details.
+
+    Returns
+    -------
+    cyc : cycler
+    '''
+    phi = 137.508 * np.pi / 180.
+
+    half_x = x_range / 2
+    half_y = y_range / 2
+
+    x_points, y_points = [], []
+
+    diag = np.sqrt(half_x ** 2 + half_y ** 2)
+    num_rings = int((1.5 * diag / (dr / factor)) ** 2)
+    for i_ring in range(1, num_rings):
+        radius = np.sqrt(i_ring) * dr / factor
+        angle = phi * i_ring
+        x = radius * np.cos(angle)
+        y = radius * np.sin(angle)
+
+        if abs(x) <= half_x and abs(y) <= half_y:
+            x_points.append(x_start + x)
+            y_points.append(y_start + y)
+
+    cyc = cycler(x_motor, x_points)
+    cyc += cycler(y_motor, y_points)
+    return cyc
+
+
+def inner_product(num, args):
+    '''Scan over one multi-motor trajectory.
+
+    Parameters
+    ----------
+    num : integer
+        number of steps
+    args : list of {Positioner, Positioner, int}
+        patterned like (``motor1, start1, stop1, ..., motorN, startN, stopN``)
+        Motors can be any 'setable' object (motor, temp controller, etc.)
+
+    Returns
+    -------
+    cyc : cycler
+    '''
+    if len(args) % 3 != 0:
+        raise ValueError("wrong number of positional arguments")
+
+    cyclers = []
+    for motor, start, stop, in chunked(args, 3):
+        steps = np.linspace(start, stop, num=num, endpoint=True)
+        c = cycler(motor, steps)
+        cyclers.append(c)
+    return functools.reduce(operator.add, cyclers)
+
+
+def chunk_outer_product_args(args):
+    '''Scan over a mesh; each motor is on an independent trajectory.
+
+    Parameters
+    ----------
+    args
+        patterned like (``motor1, start1, stop1, num1,```
+                        ``motor2, start2, stop2, num2, snake2,``
+                        ``motor3, start3, stop3, num3, snake3,`` ...
+                        ``motorN, startN, stopN, numN, snakeN``)
+
+        The first motor is the "slowest", the outer loop. For all motors
+        except the first motor, there is a "snake" argument: a boolean
+        indicating whether to following snake-like, winding trajectory or a
+        simple left-to-right trajectory.
+
+    See Also
+    --------
+    `bluesky.plan_patterns.outer_product`
+
+    Yields
+    ------
+    (motor, start, stop, num, snake)
+    '''
+    args = list(args)
+    # The first (slowest) axis is never "snaked." Insert False to
+    # make it easy to iterate over the chunks or args..
+    args.insert(4, False)
+    if len(args) % 5 != 0:
+        raise ValueError("wrong number of positional arguments")
+
+    yield from chunked(args, 5)
```

Comment:
As much as I love generators, I think I would rather `return list(chunked(args, 5))` for consistency with the rest of these.